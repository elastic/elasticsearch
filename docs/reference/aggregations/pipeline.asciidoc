[[search-aggregations-pipeline]]

== Pipeline aggregations

Pipeline aggregations work on the outputs produced from other aggregations rather than from document sets, adding
information to the output tree. There are many different types of pipeline aggregation, each computing different information from
other aggregations, but these types can be broken down into two families:

_Parent_::
                A family of pipeline aggregations that is provided with the output of its parent aggregation and is able
                to compute new buckets or new aggregations to add to existing buckets.

_Sibling_::
                Pipeline aggregations that are provided with the output of a sibling aggregation and are able to compute a
                new aggregation which will be at the same level as the sibling aggregation.

Pipeline aggregations can reference the aggregations they need to perform their computation by using the `buckets_path`
parameter to indicate the paths to the required metrics. The syntax for defining these paths can be found in the
<<buckets-path-syntax, `buckets_path` Syntax>> section below.

Pipeline aggregations cannot have sub-aggregations but depending on the type it can reference another pipeline in the `buckets_path`
allowing pipeline aggregations to be chained. For example, you can chain together two derivatives to calculate the second derivative
(i.e. a derivative of a derivative).

NOTE: Because pipeline aggregations only add to the output, when chaining pipeline aggregations the output of each pipeline aggregation
will be included in the final output.

[[buckets-path-syntax]]
[discrete]
=== `buckets_path` Syntax

Most pipeline aggregations require another aggregation as their input. The input aggregation is defined via the `buckets_path`
parameter, which follows a specific format:

// https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_Form
[source,ebnf]
--------------------------------------------------
AGG_SEPARATOR       =  `>` ;
METRIC_SEPARATOR    =  `.` ;
AGG_NAME            =  <the name of the aggregation> ;
METRIC              =  <the name of the metric (in case of multi-value metrics aggregation)> ;
MULTIBUCKET_KEY     =  `[<KEY_NAME>]`
PATH                =  <AGG_NAME><MULTIBUCKET_KEY>? (<AGG_SEPARATOR>, <AGG_NAME> )* ( <METRIC_SEPARATOR>, <METRIC> ) ;
--------------------------------------------------

For example, the path `"my_bucket>my_stats.avg"` will path to the `avg` value in the `"my_stats"` metric, which is
contained in the `"my_bucket"` bucket aggregation.

Here are some more examples:
--
 * `multi_bucket["foo"]>single_bucket>multi_metric.avg` will go to the `avg` metric in the `"multi_metric"` agg under the
    single bucket `"single_bucket"` within the `"foo"` bucket of the `"multi_bucket"` multi-bucket aggregation.
 * `agg1["foo"]._count` will get the `_count` metric for the `"foo"` bucket in the
    multi-bucket aggregation `"multi_bucket"`
--

Paths are relative from the position of the pipeline aggregation; they are not absolute paths, and the path cannot go back "up" the
aggregation tree. For example, this derivative is embedded inside a date_histogram and refers to a "sibling"
metric `"the_sum"`:

[source,console,id=buckets-path-example]
--------------------------------------------------
POST /_search
{
  "aggs": {
    "my_date_histo": {
      "date_histogram": {
        "field": "timestamp",
        "calendar_interval": "day"
      },
      "aggs": {
        "the_sum": {
          "sum": { "field": "lemmings" }              <1>
        },
        "the_deriv": {
          "derivative": { "buckets_path": "the_sum" } <2>
        }
      }
    }
  }
}
--------------------------------------------------

<1> The metric is called `"the_sum"`
<2> The `buckets_path` refers to the metric via a relative path `"the_sum"`

`buckets_path` is also used for Sibling pipeline aggregations, where the aggregation is "next" to a series of buckets
instead of embedded "inside" them. For example, the `max_bucket` aggregation uses the `buckets_path` to specify
a metric embedded inside a sibling aggregation:

[source,console,id=buckets-path-sibling-example]
--------------------------------------------------
POST /_search
{
  "aggs": {
    "sales_per_month": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "month"
      },
      "aggs": {
        "sales": {
          "sum": {
            "field": "price"
          }
        }
      }
    },
    "max_monthly_sales": {
      "max_bucket": {
        "buckets_path": "sales_per_month>sales" <1>
      }
    }
  }
}
--------------------------------------------------
// TEST[setup:sales]

<1> `buckets_path` instructs this max_bucket aggregation that we want the maximum value of the `sales` aggregation in the
`sales_per_month` date histogram.

If a Sibling pipeline agg references a multi-bucket aggregation, such as a `terms` agg, it also has the option to
select specific keys from the multi-bucket. For example, a `bucket_script` could select two specific buckets (via
their bucket keys) to perform the calculation:

[source,console,id=buckets-path-specific-bucket-example]
--------------------------------------------------
POST /_search
{
  "aggs": {
    "sales_per_month": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "month"
      },
      "aggs": {
        "sale_type": {
          "terms": {
            "field": "type"
          },
          "aggs": {
            "sales": {
              "sum": {
                "field": "price"
              }
            }
          }
        },
        "hat_vs_bag_ratio": {
          "bucket_script": {
            "buckets_path": {
              "hats": "sale_type['hat']>sales",   <1>
              "bags": "sale_type['bag']>sales"    <1>
            },
            "script": "params.hats / params.bags"
          }
        }
      }
    }
  }
}
--------------------------------------------------
// TEST[setup:sales]

<1> `buckets_path` selects the hats and bags buckets (via `['hat']`/`['bag']``) to use in the script specifically,
instead of fetching all the buckets from `sale_type` aggregation

[discrete]
=== Special Paths

Instead of pathing to a metric, `buckets_path` can use a special `"_count"` path. This instructs
the pipeline aggregation to use the document count as its input. For example, a derivative can be calculated
on the document count of each bucket, instead of a specific metric:

[source,console,id=buckets-path-count-example]
--------------------------------------------------
POST /_search
{
  "aggs": {
    "my_date_histo": {
      "date_histogram": {
        "field": "timestamp",
        "calendar_interval": "day"
      },
      "aggs": {
        "the_deriv": {
          "derivative": { "buckets_path": "_count" } <1>
        }
      }
    }
  }
}
--------------------------------------------------

<1> By using `_count` instead of a metric name, we can calculate the derivative of document counts in the histogram

The `buckets_path` can also use `"_bucket_count"` and path to a multi-bucket aggregation to use the number of buckets
returned by that aggregation in the pipeline aggregation instead of a metric. For example, a `bucket_selector` can be
used here to filter out buckets which contain no buckets for an inner terms aggregation:

[source,console,id=buckets-path-bucket-count-example]
--------------------------------------------------
POST /sales/_search
{
  "size": 0,
  "aggs": {
    "histo": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "day"
      },
      "aggs": {
        "categories": {
          "terms": {
            "field": "category"
          }
        },
        "min_bucket_selector": {
          "bucket_selector": {
            "buckets_path": {
              "count": "categories._bucket_count" <1>
            },
            "script": {
              "source": "params.count != 0"
            }
          }
        }
      }
    }
  }
}
--------------------------------------------------
// TEST[setup:sales]

<1> By using `_bucket_count` instead of a metric name, we can filter out `histo` buckets where they contain no buckets
for the `categories` aggregation

[[dots-in-agg-names]]
[discrete]
=== Dealing with dots in agg names

An alternate syntax is supported to cope with aggregations or metrics which
have dots in the name, such as the ++99.9++th
<<search-aggregations-metrics-percentile-aggregation,percentile>>. This metric
may be referred to as:

[source,js]
---------------
"buckets_path": "my_percentile[99.9]"
---------------
// NOTCONSOLE

[[gap-policy]]
[discrete]
=== Dealing with gaps in the data

Data in the real world is often noisy and sometimes contains *gaps* -- places where data simply doesn't exist. This can
occur for a variety of reasons, the most common being:

* Documents falling into a bucket do not contain a required field
* There are no documents matching the query for one or more buckets
* The metric being calculated is unable to generate a value, likely because another dependent bucket is missing a value.
Some pipeline aggregations have specific requirements that must be met (e.g. a derivative cannot calculate a metric for the
first value because there is no previous value, HoltWinters moving average need "warmup" data to begin calculating, etc)

Gap policies are a mechanism to inform the pipeline aggregation about the desired behavior when "gappy" or missing
data is encountered. All pipeline aggregations accept the `gap_policy` parameter. There are currently two gap policies
to choose from:

_skip_::
                This option treats missing data as if the bucket does not exist. It will skip the bucket and continue
                calculating using the next available value.

_insert_zeros_::
                This option will replace missing values with a zero (`0`) and pipeline aggregation computation will
                proceed as normal.

_keep_values_::
                This option is similar to skip, except if the metric provides a non-null, non-NaN value this value is
                used, otherwise the empty bucket is skipped.

include::pipeline/avg-bucket-aggregation.asciidoc[]

include::pipeline/bucket-script-aggregation.asciidoc[]

include::pipeline/bucket-count-ks-test-aggregation.asciidoc[]

include::pipeline/bucket-correlation-aggregation.asciidoc[]

include::pipeline/bucket-selector-aggregation.asciidoc[]

include::pipeline/bucket-sort-aggregation.asciidoc[]

include::pipeline/change-point-aggregation.asciidoc[]

include::pipeline/cumulative-cardinality-aggregation.asciidoc[]

include::pipeline/cumulative-sum-aggregation.asciidoc[]

include::pipeline/derivative-aggregation.asciidoc[]

include::pipeline/extended-stats-bucket-aggregation.asciidoc[]

include::pipeline/inference-bucket-aggregation.asciidoc[]

include::pipeline/max-bucket-aggregation.asciidoc[]

include::pipeline/min-bucket-aggregation.asciidoc[]

include::pipeline/movfn-aggregation.asciidoc[]

include::pipeline/moving-percentiles-aggregation.asciidoc[]

include::pipeline/normalize-aggregation.asciidoc[]

include::pipeline/percentiles-bucket-aggregation.asciidoc[]

include::pipeline/serial-diff-aggregation.asciidoc[]

include::pipeline/stats-bucket-aggregation.asciidoc[]

include::pipeline/sum-bucket-aggregation.asciidoc[]
