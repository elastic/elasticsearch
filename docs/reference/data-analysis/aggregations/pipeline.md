---
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline.html
---

# Pipeline [search-aggregations-pipeline]

Pipeline aggregations work on the outputs produced from other aggregations rather than from document sets, adding information to the output tree. There are many different types of pipeline aggregation, each computing different information from other aggregations, but these types can be broken down into two families:

*Parent*
:   A family of pipeline aggregations that is provided with the output of its parent aggregation and is able to compute new buckets or new aggregations to add to existing buckets.

*Sibling*
:   Pipeline aggregations that are provided with the output of a sibling aggregation and are able to compute a new aggregation which will be at the same level as the sibling aggregation.

Pipeline aggregations can reference the aggregations they need to perform their computation by using the `buckets_path` parameter to indicate the paths to the required metrics. The syntax for defining these paths can be found in the [`buckets_path` Syntax](#buckets-path-syntax) section below.

Pipeline aggregations cannot have sub-aggregations but depending on the type it can reference another pipeline in the `buckets_path` allowing pipeline aggregations to be chained. For example, you can chain together two derivatives to calculate the second derivative (i.e. a derivative of a derivative).

::::{note}
Because pipeline aggregations only add to the output, when chaining pipeline aggregations the output of each pipeline aggregation will be included in the final output.
::::



## `buckets_path` Syntax [buckets-path-syntax]

Most pipeline aggregations require another aggregation as their input. The input aggregation is defined via the `buckets_path` parameter, which follows a specific format:

```ebnf
AGG_SEPARATOR       =  `>` ;
METRIC_SEPARATOR    =  `.` ;
AGG_NAME            =  <the name of the aggregation> ;
METRIC              =  <the name of the metric (in case of multi-value metrics aggregation)> ;
MULTIBUCKET_KEY     =  `[<KEY_NAME>]`
PATH                =  <AGG_NAME><MULTIBUCKET_KEY>? (<AGG_SEPARATOR>, <AGG_NAME> )* ( <METRIC_SEPARATOR>, <METRIC> ) ;
```

For example, the path `"my_bucket>my_stats.avg"` will path to the `avg` value in the `"my_stats"` metric, which is contained in the `"my_bucket"` bucket aggregation.

Here are some more examples:

* `multi_bucket["foo"]>single_bucket>multi_metric.avg` will go to the `avg` metric in the `"multi_metric"` agg under the single bucket `"single_bucket"` within the `"foo"` bucket of the `"multi_bucket"` multi-bucket aggregation.
* `agg1["foo"]._count` will get the `_count` metric for the `"foo"` bucket in the multi-bucket aggregation `"multi_bucket"`

Paths are relative from the position of the pipeline aggregation; they are not absolute paths, and the path cannot go back "up" the aggregation tree. For example, this derivative is embedded inside a date_histogram and refers to a "sibling" metric `"the_sum"`:

$$$buckets-path-example$$$

```console
POST /_search
{
  "aggs": {
    "my_date_histo": {
      "date_histogram": {
        "field": "timestamp",
        "calendar_interval": "day"
      },
      "aggs": {
        "the_sum": {
          "sum": { "field": "lemmings" }              <1>
        },
        "the_deriv": {
          "derivative": { "buckets_path": "the_sum" } <2>
        }
      }
    }
  }
}
```

1. The metric is called `"the_sum"`
2. The `buckets_path` refers to the metric via a relative path `"the_sum"`


`buckets_path` is also used for Sibling pipeline aggregations, where the aggregation is "next" to a series of buckets instead of embedded "inside" them. For example, the `max_bucket` aggregation uses the `buckets_path` to specify a metric embedded inside a sibling aggregation:

$$$buckets-path-sibling-example$$$

```console
POST /_search
{
  "aggs": {
    "sales_per_month": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "month"
      },
      "aggs": {
        "sales": {
          "sum": {
            "field": "price"
          }
        }
      }
    },
    "max_monthly_sales": {
      "max_bucket": {
        "buckets_path": "sales_per_month>sales" <1>
      }
    }
  }
}
```

1. `buckets_path` instructs this max_bucket aggregation that we want the maximum value of the `sales` aggregation in the `sales_per_month` date histogram.


If a Sibling pipeline agg references a multi-bucket aggregation, such as a `terms` agg, it also has the option to select specific keys from the multi-bucket. For example, a `bucket_script` could select two specific buckets (via their bucket keys) to perform the calculation:

$$$buckets-path-specific-bucket-example$$$

```console
POST /_search
{
  "aggs": {
    "sales_per_month": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "month"
      },
      "aggs": {
        "sale_type": {
          "terms": {
            "field": "type"
          },
          "aggs": {
            "sales": {
              "sum": {
                "field": "price"
              }
            }
          }
        },
        "hat_vs_bag_ratio": {
          "bucket_script": {
            "buckets_path": {
              "hats": "sale_type['hat']>sales",   <1>
              "bags": "sale_type['bag']>sales"    <1>
            },
            "script": "params.hats / params.bags"
          }
        }
      }
    }
  }
}
```

1. `buckets_path` selects the hats and bags buckets (via `['hat']`/`['bag']``) to use in the script specifically, instead of fetching all the buckets from `sale_type` aggregation



## Special Paths [_special_paths]

Instead of pathing to a metric, `buckets_path` can use a special `"_count"` path. This instructs the pipeline aggregation to use the document count as its input. For example, a derivative can be calculated on the document count of each bucket, instead of a specific metric:

$$$buckets-path-count-example$$$

```console
POST /_search
{
  "aggs": {
    "my_date_histo": {
      "date_histogram": {
        "field": "timestamp",
        "calendar_interval": "day"
      },
      "aggs": {
        "the_deriv": {
          "derivative": { "buckets_path": "_count" } <1>
        }
      }
    }
  }
}
```

1. By using `_count` instead of a metric name, we can calculate the derivative of document counts in the histogram


The `buckets_path` can also use `"_bucket_count"` and path to a multi-bucket aggregation to use the number of buckets returned by that aggregation in the pipeline aggregation instead of a metric. For example, a `bucket_selector` can be used here to filter out buckets which contain no buckets for an inner terms aggregation:

$$$buckets-path-bucket-count-example$$$

```console
POST /sales/_search
{
  "size": 0,
  "aggs": {
    "histo": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "day"
      },
      "aggs": {
        "categories": {
          "terms": {
            "field": "category"
          }
        },
        "min_bucket_selector": {
          "bucket_selector": {
            "buckets_path": {
              "count": "categories._bucket_count" <1>
            },
            "script": {
              "source": "params.count != 0"
            }
          }
        }
      }
    }
  }
}
```

1. By using `_bucket_count` instead of a metric name, we can filter out `histo` buckets where they contain no buckets for the `categories` aggregation



## Dealing with dots in agg names [dots-in-agg-names]

An alternate syntax is supported to cope with aggregations or metrics which have dots in the name, such as the `99.9`th [percentile](/reference/data-analysis/aggregations/search-aggregations-metrics-percentile-aggregation.md). This metric may be referred to as:

```js
"buckets_path": "my_percentile[99.9]"
```


## Dealing with gaps in the data [gap-policy]

Data in the real world is often noisy and sometimes contains **gaps** — places where data simply doesn’t exist. This can occur for a variety of reasons, the most common being:

* Documents falling into a bucket do not contain a required field
* There are no documents matching the query for one or more buckets
* The metric being calculated is unable to generate a value, likely because another dependent bucket is missing a value. Some pipeline aggregations have specific requirements that must be met (e.g. a derivative cannot calculate a metric for the first value because there is no previous value, HoltWinters moving average need "warmup" data to begin calculating, etc)

Gap policies are a mechanism to inform the pipeline aggregation about the desired behavior when "gappy" or missing data is encountered. All pipeline aggregations accept the `gap_policy` parameter. There are currently two gap policies to choose from:

*skip*
:   This option treats missing data as if the bucket does not exist. It will skip the bucket and continue calculating using the next available value.

*insert_zeros*
:   This option will replace missing values with a zero (`0`) and pipeline aggregation computation will proceed as normal.

*keep_values*
:   This option is similar to skip, except if the metric provides a non-null, non-NaN value this value is used, otherwise the empty bucket is skipped.






















