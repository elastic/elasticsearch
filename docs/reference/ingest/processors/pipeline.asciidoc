[[pipeline-processor]]
=== Pipeline processor
++++
<titleabbrev>Pipeline</titleabbrev>
++++

Executes another pipeline.

[[pipeline-options]]
.Pipeline Options
[options="header"]
|======
| Name                      | Required  | Default  | Description
| `name`                    | yes       | -        | The name of the pipeline to execute. Supports <<template-snippets,template snippets>>.
| `ignore_missing_pipeline` | no        | false    | Whether to ignore missing pipelines instead of failing.
include::common-options.asciidoc[]
|======

[source,js]
--------------------------------------------------
{
  "pipeline": {
    "name": "inner-pipeline"
  }
}
--------------------------------------------------
// NOTCONSOLE

The name of the current pipeline can be accessed from the `_ingest.pipeline` ingest metadata key.

An example of using this processor for nesting pipelines would be:

Define an inner pipeline:

[source,console]
--------------------------------------------------
PUT _ingest/pipeline/pipelineA
{
  "description" : "inner pipeline",
  "processors" : [
    {
      "set" : {
        "field": "inner_pipeline_set",
        "value": "inner"
      }
    }
  ]
}
--------------------------------------------------

Define another pipeline that uses the previously defined inner pipeline:

[source,console]
--------------------------------------------------
PUT _ingest/pipeline/pipelineB
{
  "description" : "outer pipeline",
  "processors" : [
    {
      "pipeline" : {
        "name": "pipelineA"
      }
    },
    {
      "set" : {
        "field": "outer_pipeline_set",
        "value": "outer"
      }
    }
  ]
}
--------------------------------------------------
// TEST[continued]

Now indexing a document while applying the outer pipeline will see the inner pipeline executed
from the outer pipeline:

[source,console]
--------------------------------------------------
PUT /my-index-000001/_doc/1?pipeline=pipelineB
{
  "field": "value"
}
--------------------------------------------------
// TEST[continued]

Response from the index request:

[source,console-result]
--------------------------------------------------
{
  "_index": "my-index-000001",
  "_id": "1",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 66,
  "_primary_term": 1
}
--------------------------------------------------
// TESTRESPONSE[s/"_seq_no": \d+/"_seq_no" : $body._seq_no/ s/"_primary_term" : 1/"_primary_term" : $body._primary_term/]

Indexed document:

[source,js]
--------------------------------------------------
{
  "field": "value",
  "inner_pipeline_set": "inner",
  "outer_pipeline_set": "outer"
}
--------------------------------------------------
// NOTCONSOLE
