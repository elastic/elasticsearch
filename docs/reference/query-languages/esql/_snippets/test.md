This is a test

:::{include} lists/aggregation-functions.md
:::
