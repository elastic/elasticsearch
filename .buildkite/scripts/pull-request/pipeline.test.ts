import { beforeEach, describe, expect, test } from "bun:test";

import { generatePipelines } from "./pipeline";
import { setBwcVersionsPath, setSnapshotBwcVersionsPath } from "./bwc-versions";

describe("generatePipelines", () => {
  beforeEach(() => {
    setBwcVersionsPath(`${import.meta.dir}/mocks/bwcVersions`);
    setSnapshotBwcVersionsPath(`${import.meta.dir}/mocks/snapshotBwcVersions`);

    process.env["GITHUB_PR_LABELS"] = "test-label-1,test-label-2";
    process.env["GITHUB_PR_TRIGGER_COMMENT"] = "";
  });

  // Helper for testing pipeline generations that should be the same when using the overall ci trigger comment "buildkite test this"
  const testWithTriggerCheck = (directory: string, changedFiles?: string[], comment = "buildkite test this") => {
    const pipelines = generatePipelines(directory, changedFiles);
    expect(pipelines).toMatchSnapshot();

    process.env["GITHUB_PR_TRIGGER_COMMENT"] = comment;
    const pipelinesWithTriggerComment = generatePipelines(directory, changedFiles);
    expect(pipelinesWithTriggerComment).toEqual(pipelines);
  };

  test("should generate correct pipelines with a non-docs change", () => {
    testWithTriggerCheck(`${import.meta.dir}/mocks/pipelines`, ["build.gradle", "docs/README.asciidoc"]);
  });

  test("should generate correct pipelines with only docs changes", () => {
    testWithTriggerCheck(`${import.meta.dir}/mocks/pipelines`, ["docs/README.asciidoc"]);
  });

  test("should generate correct pipelines with full BWC expansion", () => {
    process.env["GITHUB_PR_LABELS"] = "test-full-bwc";

    testWithTriggerCheck(`${import.meta.dir}/mocks/pipelines`, ["build.gradle"]);
  });

  test("should generate correct pipeline when using a trigger comment for it", () => {
    process.env["GITHUB_PR_TRIGGER_COMMENT"] = "run elasticsearch-ci/using-defaults";

    const pipelines = generatePipelines(`${import.meta.dir}/mocks/pipelines`, ["build.gradle"]);
    expect(pipelines).toMatchSnapshot();
  });

  test("should generate correct pipelines with a non-docs change and @elasticmachine", () => {
    testWithTriggerCheck(
      `${import.meta.dir}/mocks/pipelines`,
      ["build.gradle", "docs/README.asciidoc"],
      "@elasticmachine test this please"
    );
  });

  test("should generate correct pipelines with a non-docs change and @elasticsearchmachine", () => {
    testWithTriggerCheck(
      `${import.meta.dir}/mocks/pipelines`,
      ["build.gradle", "docs/README.asciidoc"],
      "@elasticsearchmachine test this please"
    );
  });
});
