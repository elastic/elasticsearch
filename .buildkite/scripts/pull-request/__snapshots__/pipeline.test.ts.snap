// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePipelines should generate correct pipelines with a non-docs change 1`] = `
[
  {
    "name": "bwc-snapshots",
    "pipeline": {
      "steps": [
        {
          "group": "bwc-snapshots",
          "steps": [
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dignore.tests.seed v{{matrix.BWC_VERSION}}#bwcTest",
              "env": {
                "BWC_VERSION": "{{matrix.BWC_VERSION}}",
              },
              "label": "{{matrix.BWC_VERSION}} / bwc-snapshots",
              "matrix": {
                "setup": {
                  "BWC_VERSION": [
                    "7.17.14",
                    "8.10.3",
                    "8.11.0",
                  ],
                },
              },
              "timeout_in_minutes": 300,
            },
          ],
        },
      ],
    },
  },
  {
    "name": "using-defaults",
    "pipeline": {
      "env": {
        "CUSTOM_ENV_VAR": "value",
      },
      "steps": [
        {
          "command": "echo 'hello world'",
          "label": "test-step",
        },
      ],
    },
  },
]
`;

exports[`generatePipelines should generate correct pipelines with only docs changes 1`] = `
[
  {
    "name": "docs-check",
    "pipeline": {
      "steps": [
        {
          "agents": {
            "buildDirectory": "/dev/shm/bk",
            "image": "family/elasticsearch-ubuntu-2004",
            "machineType": "custom-32-98304",
            "provider": "gcp",
          },
          "command": ".ci/scripts/run-gradle.sh -Dignore.tests.seed precommit :docs:check",
          "label": "docs-check",
          "timeout_in_minutes": 300,
        },
      ],
    },
  },
]
`;

exports[`generatePipelines should generate correct pipelines with full BWC expansion 1`] = `
[
  {
    "name": "full-bwc",
    "pipeline": {
      "steps": [
        {
          "group": "bwc",
          "steps": [
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dbwc.checkout.align=true v7.0.0#bwcTest",
              "env": {
                "BWC_VERSION": "7.0.0",
              },
              "key": "full-bwc:7_0_0",
              "label": "7.0.0 / bwc",
              "timeout_in_minutes": 300,
            },
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dbwc.checkout.align=true v7.0.1#bwcTest",
              "env": {
                "BWC_VERSION": "7.0.1",
              },
              "key": "full-bwc:7_0_1",
              "label": "7.0.1 / bwc",
              "timeout_in_minutes": 300,
            },
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dbwc.checkout.align=true v7.1.0#bwcTest",
              "env": {
                "BWC_VERSION": "7.1.0",
              },
              "key": "full-bwc:7_1_0",
              "label": "7.1.0 / bwc",
              "timeout_in_minutes": 300,
            },
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dbwc.checkout.align=true v8.10.0#bwcTest",
              "env": {
                "BWC_VERSION": "8.10.0",
              },
              "key": "full-bwc:8_10_0",
              "label": "8.10.0 / bwc",
              "timeout_in_minutes": 300,
            },
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dbwc.checkout.align=true v8.11.0#bwcTest",
              "env": {
                "BWC_VERSION": "8.11.0",
              },
              "key": "full-bwc:8_11_0",
              "label": "8.11.0 / bwc",
              "timeout_in_minutes": 300,
            },
          ],
        },
      ],
    },
  },
  {
    "name": "using-defaults",
    "pipeline": {
      "env": {
        "CUSTOM_ENV_VAR": "value",
      },
      "steps": [
        {
          "command": "echo 'hello world'",
          "label": "test-step",
        },
      ],
    },
  },
]
`;

exports[`generatePipelines should generate correct pipeline when using a trigger comment for it 1`] = `
[
  {
    "name": "using-defaults",
    "pipeline": {
      "env": {
        "CUSTOM_ENV_VAR": "value",
      },
      "steps": [
        {
          "command": "echo 'hello world'",
          "label": "test-step",
        },
      ],
    },
  },
]
`;

exports[`generatePipelines should generate correct pipelines with a non-docs change and @elasticmachine 1`] = `
[
  {
    "name": "bwc-snapshots",
    "pipeline": {
      "steps": [
        {
          "group": "bwc-snapshots",
          "steps": [
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dignore.tests.seed v{{matrix.BWC_VERSION}}#bwcTest",
              "env": {
                "BWC_VERSION": "{{matrix.BWC_VERSION}}",
              },
              "label": "{{matrix.BWC_VERSION}} / bwc-snapshots",
              "matrix": {
                "setup": {
                  "BWC_VERSION": [
                    "7.17.14",
                    "8.10.3",
                    "8.11.0",
                  ],
                },
              },
              "timeout_in_minutes": 300,
            },
          ],
        },
      ],
    },
  },
  {
    "name": "using-defaults",
    "pipeline": {
      "env": {
        "CUSTOM_ENV_VAR": "value",
      },
      "steps": [
        {
          "command": "echo 'hello world'",
          "label": "test-step",
        },
      ],
    },
  },
]
`;

exports[`generatePipelines should generate correct pipelines with a non-docs change and @elasticsearchmachine 1`] = `
[
  {
    "name": "bwc-snapshots",
    "pipeline": {
      "steps": [
        {
          "group": "bwc-snapshots",
          "steps": [
            {
              "agents": {
                "buildDirectory": "/dev/shm/bk",
                "image": "family/elasticsearch-ubuntu-2004",
                "machineType": "custom-32-98304",
                "provider": "gcp",
              },
              "command": ".ci/scripts/run-gradle.sh -Dignore.tests.seed v{{matrix.BWC_VERSION}}#bwcTest",
              "env": {
                "BWC_VERSION": "{{matrix.BWC_VERSION}}",
              },
              "label": "{{matrix.BWC_VERSION}} / bwc-snapshots",
              "matrix": {
                "setup": {
                  "BWC_VERSION": [
                    "7.17.14",
                    "8.10.3",
                    "8.11.0",
                  ],
                },
              },
              "timeout_in_minutes": 300,
            },
          ],
        },
      ],
    },
  },
  {
    "name": "using-defaults",
    "pipeline": {
      "env": {
        "CUSTOM_ENV_VAR": "value",
      },
      "steps": [
        {
          "command": "echo 'hello world'",
          "label": "test-step",
        },
      ],
    },
  },
]
`;
